clang/lib/CIR/CodeGen/CIRGenDecl.cpp

```
@@ -0,0 +1,82 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This contains code to emit Decl nodes as CIR code.
+//
+//===----------------------------------------------------------------------===//
+
+#include "CIRGenFunction.h"
+#include "clang/AST/Expr.h"
+#include "clang/CIR/MissingFeatures.h"
+
+using namespace clang;
+using namespace clang::CIRGen;
+
+/// Emit code and set up symbol table for a variable declaration with auto,
+/// register, or no storage class specifier. These turn into simple stack
+/// objects, globals depending on target.
+void CIRGenFunction::emitAutoVarDecl(const VarDecl &d) {
+  QualType ty = d.getType();
+  assert(ty.getAddressSpace() == LangAS::Default);
```

Is this a 'not yet implemented' sorta thing?

---

Guidance here was to use the errorNYI instead of 'unreachable', as 'unreachable' is a much less 'nice' user interface.  So we leave `unreachable` for things that are NEVER going to be reachable.

---

OK. I like `unreachable` because the diagnostic points me to the place in the code where the implementation is missing rather than the place in the user source, which is probably a test case that I just added. But I'm happy to go along with the consensus.

---

> upstreaming seems to be a mix of errorNYI and silently ignoring things

If upstream is silently ignoring things this is pretty bad, because it leads to subtle miscompilations that are super hard to debug/find. My understanding is that `errorNYI` should be covering all cases where we use unrecheable in the incubator, but if that isn't enough, unrecheable is 1000x better than silently ignoring

---

Ideally, every not-yet-implemented or not-yet-upstreamed language construct should result in a user error, which is most easily done by calling `errorNYI`.  In the incubator, those situations usually have an `llvm_unreachable("NYI")` or an `assert(condition && "NYI")`.  We would prefer that upstream use `errorNYI` instead.  If there is code where reporting a user error is not feasible (I am not familiar enough with Clang or MLIR to know how often that will happen), then an `llvm_unreachable("NYI: <unimplemented feature>")` will have to do.

It would be great if every unimplemented language construct was checked for and resulted in an error, but I don't think that's feasible at the moment.  The number of unimplemented attributes and attribute-like properties that can be attached to declarations is huge, and adding a check for every one of them would be tedious and would overwhelm the happy-path code.  As long as the upstreamed ClangIR is unable to compile anything useful, I think this is okay.  Right now no one will compile real code and be puzzled at some subtle behavior change due to an ignored attribute.  Adding all the checks for unimplemented attributes will be important when upstreamed ClangIR is able to compile real code, and by that time many of the attributes will already be implemented and won't need a temporary NYI check.

---

> If upstream is silently ignoring things this is pretty bad, because it leads to subtle miscompilations that are super hard to debug/find.

There's nothing really subtle about no CIR being generated for function arguments (the case I had in mind that's being ignored currently). I take your point though, and I'll try to incorporate the suggestion.

My question is more like this. When we have large sections of code being omitted that are checking for conditions that maybe can't even be checked yet because the support functions are missing, how should that be handled? In this function, I have checks plus errorNYI for a lot of things, but it feels very messy. What I really want is something like a single check for the limited cases we are handling, and an error for anything else, but that's not always possible. In places where the conditions can't be checked, I've put MissingFeature calls.

---

> In places where the conditions can't be checked, I've put MissingFeature calls.

Gotcha! Yes, I think this is the best we can do in those scenarios!